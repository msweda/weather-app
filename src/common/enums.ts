export enum RootRoute {
  Home = '/',
  Weather = '/weather',
  Forecast = '/forecast',
}

export enum AsyncStatus {
  Idle = 'IDLE',
  InProgress = 'IN_PROGRESS',
  Succeeded = 'SUCCEEDED',
  Failed = 'FAILED',
}
