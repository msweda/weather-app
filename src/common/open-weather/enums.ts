export enum UnitType {
  Scientific = 'scientific',
  Metric = 'metric',
  Imperial = 'imperial',
}
