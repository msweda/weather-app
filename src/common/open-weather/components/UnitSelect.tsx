import { UnitType } from '@/common/open-weather/enums'
import { AllStores } from '@/src/stores'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import {
  createStyles,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core/styles'
import { inject, observer } from 'mobx-react'
import * as React from 'react'
import OpenWeatherStore from '../store'

@inject((allStores: AllStores) => ({
  openWeatherStore: allStores.openWeatherStore,
}))
@observer
class UnitSelect extends React.Component<Props> {
  get injected() {
    return this.props as InjectedProps
  }

  render() {
    const { classes } = this.props
    return (
      <FormControl className={classes.selectFormControl}>
        <InputLabel htmlFor="weather-units-input">Units</InputLabel>
        <Select
          className={classes.select}
          value={this.injected.openWeatherStore.unitType}
          onChange={this.changeUnitType}
          inputProps={{ id: 'weather-units-input' }}
        >
          <MenuItem className={classes.menuItem} value={UnitType.Scientific}>
            Kelvin
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Metric}>
            Celsius
          </MenuItem>
          <MenuItem className={classes.menuItem} value={UnitType.Imperial}>
            Fahrenheit
          </MenuItem>
        </Select>
      </FormControl>
    )
  }

  private changeUnitType = (event: React.ChangeEvent<HTMLSelectElement>) => {
    this.injected.openWeatherStore.setUnitType(event.target.value as UnitType)
    this.injected.openWeatherStore.clearWeather()
    this.injected.openWeatherStore.clearForecast()
  }
}

const styles = (theme: Theme) =>
  createStyles({
    selectFormControl: {
      marginBottom: 20,
    },
    select: {
      color: theme.palette.primary.main,
    },
    menuItem: {
      color: theme.palette.primary.main,
    },
  })

interface Props extends WithStyles<typeof styles> {}

interface InjectedProps extends Props {
  openWeatherStore: OpenWeatherStore
}

export default withStyles(styles)(UnitSelect)
