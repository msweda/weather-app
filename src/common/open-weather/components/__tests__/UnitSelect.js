import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import { shallow } from 'enzyme'
import * as React from 'react'
import { UnitType } from '../../enums'
import OpenWeatherStore from '../../store'
import UnitSelect from '../UnitSelect'

jest.mock('../../store')

const openWeatherStore = new OpenWeatherStore()

describe('<UnitSelect />', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <UnitSelect.wrappedComponent
        openWeatherStore={openWeatherStore}
        classes={{}}
      />
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('provides an option for Kelvin', () => {
    const menuItem = wrapper.find(MenuItem).at(0)
    expect(menuItem.children().text()).toEqual('Kelvin')
  })

  it('provides an option for Celsius', () => {
    const menuItem = wrapper.find(MenuItem).at(1)
    expect(menuItem.children().text()).toEqual('Celsius')
  })

  it('provides an option for Fahrenheit', () => {
    const menuItem = wrapper.find(MenuItem).at(2)
    expect(menuItem.children().text()).toEqual('Fahrenheit')
  })

  it('informs the store to update its unit type on changes', () => {
    wrapper
      .find(Select)
      .simulate('change', { target: { value: UnitType.Metric } })
    expect(openWeatherStore.setUnitType).toHaveBeenCalledWith(UnitType.Metric)
    wrapper
      .find(Select)
      .simulate('change', { target: { value: UnitType.Imperial } })
    expect(openWeatherStore.setUnitType).toHaveBeenCalledWith(UnitType.Imperial)
    wrapper
      .find(Select)
      .simulate('change', { target: { value: UnitType.Scientific } })
    expect(openWeatherStore.setUnitType).toHaveBeenCalledWith(
      UnitType.Scientific
    )
  })

  it('informs the store to clear its weather and forecast data on change', () => {
    wrapper
      .find(Select)
      .simulate('change', { target: { value: UnitType.Metric } })
    expect(openWeatherStore.clearWeather).toHaveBeenCalled()
    expect(openWeatherStore.clearForecast).toHaveBeenCalled()
  })
})
