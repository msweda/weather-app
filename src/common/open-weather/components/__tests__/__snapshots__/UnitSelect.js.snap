// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnitSelect /> should match snapshot 1`] = `
<WithStyles(FormControl)>
  <WithStyles(WithFormControlContext(InputLabel))
    htmlFor="weather-units-input"
  >
    Units
  </WithStyles(WithFormControlContext(InputLabel))>
  <WithStyles(WithFormControlContext(Select))
    inputProps={
      Object {
        "id": "weather-units-input",
      }
    }
    onChange={[Function]}
  >
    <WithStyles(MenuItem)
      value="scientific"
    >
      Kelvin
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value="metric"
    >
      Celsius
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      value="imperial"
    >
      Fahrenheit
    </WithStyles(MenuItem)>
  </WithStyles(WithFormControlContext(Select))>
</WithStyles(FormControl)>
`;
